internal/util/pool.go

```diff
@@ -6,105 +6,175 @@ import (
+func RunParallelPoolMap[InputType comparable, OutputType any](poolSize int, workItems []InputType, ctx context.Context, workFunc func(InputType) (OutputType, error)) (*PoolResult[InputType, OutputType], error) {
 	if poolSize <= 0 {
-		return nil, nil, fmt.Errorf("Pool size must be positive, got %d.", poolSize)
-	}
-
-	type WorkItem struct {
-		Index int
-		Item  InputType
+		return nil, fmt.Errorf("Pool size must be positive, got %d.", poolSize)
 	}
 
 	type ResultItem struct {
-		Index int
-		Item  OutputType
-		Error error
+		Input  InputType
+		Result OutputType
+		Error  error
 	}
 
-	results := make([]OutputType, len(workItems))
-	workErrors := make(map[int]error)
-
 	resultQueue := make(chan ResultItem, poolSize)
-	workQueue := make(chan WorkItem, poolSize)
-	doneChan := make(chan bool, poolSize)
-	exitWaitGroup := sync.WaitGroup{}
+	workQueue := make(chan InputType, poolSize)
+
+	// The done working channel is used by the pool workers to signal work is complete to the result collectors.
+	doneWorkingChan := make(chan any)
+	// The done channel is used to signal that the all results are collected and accessible.
+	// It also waits to signal until all worker threads join the main thread.
+	doneChan := make(chan any)
+
+	// A wait group used to determine when the workers have completed there work.
+	// Note that completion can be natural or due to a cancellation.
+	workerExitWaitGroup := sync.WaitGroup{}
+	workerExitWaitGroup.Add(poolSize)
+
+	// There will always be poolSize + 2 threads started.
+	// A thread will be started for the work loader, the result collector, and each worker in the pool.
+	// Every one of these work threads must call threadExitWaitGroup.Done() to avoid a deadlock.
+	threadExitWaitGroup := sync.WaitGroup{}
+	threadExitWaitGroup.Add(poolSize + 2)
+
+	output := PoolResult[InputType, OutputType]{
+		Results:    make(map[InputType]OutputType, len(workItems)),
+		WorkErrors: make(map[InputType]error, 0),
+		done:       doneChan,
+	}
 
 	// Load work.
 	go func() {
-		for i, item := range workItems {
+		defer threadExitWaitGroup.Done()
+		// Signal to the workers that there is no more work.
+		defer close(workQueue)
+
+		for _, item := range workItems {
 			// Either send a work item, or cancel.
 			select {
 			case <-ctx.Done():
 				return
-			case workQueue <- WorkItem{i, item}:
+			case workQueue <- item:
 				// Item was already sent on the chan, do nothing here.
 			}
 		}
 	}()
 
 	// Collect results.
 	go func() {
+		defer func() {
+			defer threadExitWaitGroup.Done()
+
+			// All workers completed, drain remaining results.
+			// If the context was canceled, the number of results may be less than the number of work items.
+			for {
+				select {
+				case resultItem := <-resultQueue:
+					output.Results[resultItem.Input] = resultItem.Result
+					if resultItem.Error != nil {
+						output.WorkErrors[resultItem.Input] = resultItem.Error
+					}
```

Comment:
This is inconsistent with the function-level comment:
> // every input will either be in PoolResult.Results or PoolResult.WorkErrors.

Just use if/else here and in the primary collection to make sure only one gets populated for each item.